src/test/scala/org/constellation/primitives/TipServiceTest.scala

```diff
@@ -0,0 +1,96 @@
+package org.constellation.primitives
+import cats.effect.{ContextShift, IO}
+import cats.implicits._
+import org.constellation.DAO
+import org.constellation.consensus.TipData
+import org.constellation.util.Metrics
+import org.mockito.integrations.scalatest.IdiomaticMockitoFixture
+import org.scalatest.FunSpecLike
+
+import scala.util.Try
+
+class TipServiceTest extends FunSpecLike with IdiomaticMockitoFixture {
+
+  implicit val dao: DAO = prepareDAO()
+
+  def prepareDAO(): DAO = {
+    val dao = mock[DAO]
+    dao.metrics shouldReturn mock[Metrics]
+    dao.threadSafeSnapshotService shouldReturn mock[ThreadSafeSnapshotService]
+    dao.threadSafeSnapshotService.acceptedCBSinceSnapshot shouldReturn Seq.empty
+    dao
+  }
+
+  describe("TipService") {
+
+    it("limits maximum number of tips") {
+      val limit = 6
+      val concurrentTipService = new TrieBasedTipService(limit, 10, 2, 30)
+
+      val cbs = createIndexedCBmocks(limit * 3, { i =>
+        createCBMock(i.toString)
+      })
+
+      val tasks = createShiftedTasks(cbs.toList, { cb =>
+        concurrentTipService.update(cb)
+      })
+      tasks.par.foreach(ioa => ioa.unsafeRunAsyncAndForget())
+      Thread.sleep(2000)
+      assertResult(limit)(concurrentTipService.toMap.size)
+    }
+
+    it("forbid adding conflicting tip") {
+      val limit = 6
+      val concurrentTipService = new TrieBasedTipService(limit, 10, 2, 30)
+      val txs = prepareTransactions()
+
+      val cbs = createIndexedCBmocks(limit * 3, { i =>
+        val cb = createCBMock(i.toString)
+        cb.transactions shouldReturn txs
+        cb
+      })
+
+      val tasks = createShiftedTasks(cbs.toList, { cb =>
+        concurrentTipService.update(cb)
+      })
+
+      val cbsNotInSnap = cbs.map(_.baseHash)
+      dao.threadSafeSnapshotService.acceptedCBSinceSnapshot shouldReturn cbsNotInSnap
+
+      tasks.par.foreach(ioa => ioa.unsafeRunAsyncAndForget())
+      Thread.sleep(2000)
+      assertResult(1)(concurrentTipService.toMap.size)
+    }
+  }
+
+  private def prepareTransactions(): Seq[Transaction] = {
+    val tx1 = mock[Transaction]
+    val tx2 = mock[Transaction]
+    Seq(tx1, tx2)
+  }
+
+  private def createCBMock(hash: String) = {
+    val cb = mock[CheckpointBlock]
+    cb.parentSOEBaseHashes shouldReturn Seq.empty
+    cb.transactions shouldReturn Seq.empty
+    cb.baseHash shouldReturn hash
+    cb
+  }
+
+  def createIndexedCBmocks(size: Int, func: Int => CheckpointBlock) = {
+    for (i <- 1 to size) yield {
+      func(i)
+    }
+  }
+
+  def createShiftedTasks(cbs: List[CheckpointBlock],
+                         func: CheckpointBlock => IO[Try[Option[TipData]]]) = {
+    implicit val ioContextShift: ContextShift[IO] =
```

Comment:
IMO this implicit should be above the function, and the fn itself should request for an implicit for ContextShift. this way you'll be able to override it in some specific case